lib/hooks/digest.cpp

```diff
@@ -0,0 +1,224 @@
+			throw RuntimeError { "Could not format sample into digest" };
+
+		if (!EVP_DigestUpdate(md_ctx.get(), format_buffer.data(), written))
+			throw RuntimeError { "Could not update digest" };
+	}
+
+public:
+	DigestHook(Path *p, Node *n, int fl, int prio, bool en = true) :
+		Hook(p, n, fl, prio, en),
+		algorithm(),
+		uri(),
+		mode(Mode::Timestamp),
+		interval(1.0),
+		md_ctx(EVP_MD_CTX_new(), &EVP_MD_CTX_free),
+		md(nullptr),
+		file(nullptr, &FILE_free),
+		format(),
+		current_interval(),
+		md_string_buffer(),
+		format_buffer()
+	{ }
+
+	virtual
+	void parse(json_t *json)
+	{
+		Hook::parse(json);
+
+		char const *uri_str;
+		char const *mode_str;
+		char const *algorithm_str;
+
+		json_error_t err;
+		int ret = json_unpack_ex(json, &err, 0, "{ s: s, s: s, s: s }",
+			"uri", &uri_str,
+			"mode", &mode_str,
+			"algorithm", &algorithm_str
+		);
+		if (ret)
+			throw ConfigError(json, err, "node-config-hook-digest");
+
+		uri = std::string(uri_str);
+		algorithm = std::string(algorithm_str);
+
+		if (!strcmp(mode_str, "timestamp")) {
+			mode = Mode::Timestamp;
+
+			ret = json_unpack_ex(json, &err, 0, "{ s: F }",
+				"interval", &interval.emplace<double>()
+			);
+			if (ret)
+				throw ConfigError(json, err, "node-config-hook-digest-interval");
+		} else if (!strcmp(mode_str, "sequence")) {
+			mode = Mode::Sequence;
+
+			ret = json_unpack_ex(json, &err, 0, "{ s: i }",
+				"interval", &interval.emplace<int>()
+			);
+			if (ret)
+				throw ConfigError(json, err, "node-config-hook-digest-interval");
+		}
+	}
+
+	virtual
+	void prepare()
+	{
+		Hook::prepare();
+
+		format_buffer.resize(64 * getSignals()->size());
+
+		format = format_ptr(FormatFactory::make("villas.human"));
+		if (!format)
+			throw RuntimeError { "Could not create format for digest" };
+
+		file = file_ptr(fopen(uri.c_str(), "w"), &FILE_free);
+		if (!file)
+			throw RuntimeError { "Could not open file {}: {}", uri, strerror(errno) };
+
+		md = EVP_get_digestbyname(algorithm.c_str());
+		if (!md)
+			throw RuntimeError { "Could not fetch algorithm {}", algorithm };
+	}
+
+	virtual
+	void start()
+	{
+		Hook::start();
+
+		if (!EVP_DigestInit_ex(md_ctx.get(), md, NULL))
+			throw RuntimeError { "Could not initialize digest" };
+	}
+
+	virtual
+	Hook::Reason process(struct Sample *smp)
+	{
+		assert(smp);
+		assert(state == State::STARTED);
+
+		if (updateInterval(smp)) {
+			emitDigest();
+		}
+
+		updateDigest(smp);
+
+		return Reason::OK;
+	}
+
+	virtual
+	void stop() {
+		Hook::stop();
+
+		current_interval = std::nullopt;
+		if (!EVP_MD_CTX_reset(md_ctx.get()))
+			throw RuntimeError { "Could not reset digest context" };
+	}
+};
+
+/* Register hook */
+static char n[] = "digest";
+static char d[] = "Calculate the digest for a range of samples";
+static HookPlugin<DigestHook, n, d, (int) Hook::Flags::NODE_READ | (int) Hook::Flags::NODE_WRITE | (int) Hook::Flags::PATH, 20> p;
+
+} /* namespace node */
+} /* namespace villas */
```

Comment:
I am trying to refactor the codebase to use C++ style comments (`//`) rather C-style (`/* */`)

```suggestion
} // namespace node
} // namespace villas
```